src/modules-by-example.adoc

```diff
@@ -0,0 +1,320 @@
+struct closes.
+
+A _segment_ is a contiguous portion of a document that uses the same encoding environment.
+Segment boundaries are caused by directives: an IVM starts a new segment, while an
+`$ion_symbol_table` struct ends a segment, defining a new one that starting immediately
+afterwards.
+
+TODO Ion text docs always start with a 1.0 segment until an IVM is encountered.
+
+
+=== Modules from the Outside
+
+In Ion 1.1, you define, share, and install symbols and macros using _encoding modules_.
+The logical interface to a module has three main components: a spec version, a symbol
+table, and a macro table.
+
+A module's _spec version_ indicates which Ion specification it uses. This ensures the module has
+stable semantics over time.  Spec versions are expressed as Ion version markers, and a module can
+only be used in segments encoded with that version or later.
+
+// TODO The above may be too strict; use solely for symbols could be more relaxed.
+
+A module's _exported symbol table_ is simply a sequence of strings. These denote the
+text of symbols, and are equivalent in meaning to the `symbols` list of an Ion 1.0 shared symbol
+table.
+
+A module's _exported macro table_ is a sequence of <name, macro> pairs.  Names can be
+null, in which case the corresponding macro can be referenced by its zero-based index in the
+table, known as its _exported address_.  Non-null names in the table must all be unique, so that
+a name-to-macro mapping function is well-defined.
+
+TIP: Macros have their own "object identity" independent of the names that map to them.  It's
+possible for the same macro to have multiple addresses and/or names.
+
+To reuse macros across documents, _shared modules_ subsume the capabilities
+of shared symbol tables while remaining backwards-compatible with their current schema and catalog
+semantics.
+
+IMPORTANT: Any existing Ion shared symbol table is a valid encoding module, albeit one that only
+declares symbols and not macros.
+
+
+=== Encoding Environment 1.1
+
+In Ion 1.1, the encoding environment is enhanced to include:
+
+  * The current Ion version, because a document may have segments using different Ion versions.
+  * The _available modules_, a name->module mapping.
+  * The local macro table, assembled from a subset of the available modules.
+
+Upon encountering the `$ion_1_1` IVM, the environment is reset to the default state, in which:
+
+  * The current Ion version is 1.1.
+  * The available modules contains only the `$ion` module, version 2 (v1 being Ion 1.0).
+  * The current local macro table is empty.
+  * The current symbol table is the Ion 1.1 system symbol table.
+
+To customize this environment, we use an _encoding directive_: a
+top-level struct annotated with `$ion_encoding`. Like `$ion_symbol_table`, this directive defines a
+new encoding environment that goes into effect immediately after the struct closes.
+
+The general syntax of an encoding directive is as follows:
```

Comment:
I'm using the term "encoding directive" specifically for this 1.1 entity.  `$ion_symbol_table` and IVMs are different kinds of directives.

I'm not 100% satisfied by that delineation, but it's the best I've got today.  I def want a specific term for `$ion_encoding` so that annotation isn't needed everywhere its discussed.

Reply:
I think we need some text that indicates that the LST is not an encoding directive and indicates if such a system value is allowed in Ion 1.1 or not.